coredocument/read_acls.go

```diff
@@ -110,7 +144,100 @@ func isPeerInRole(role *coredocumentpb.Role, peer identity.CentID) bool {
 	return false
 }
 
-// peerValidator return the
+// peerValidator returns the ReadAccessValidator tp verify peer.
 func peerValidator() ReadAccessValidator {
 	return readAccessValidator{}
 }
+
+// nftValidator returns the ReadAccessValidator for nft owner verification.
+func nftValidator(tr TokenRegistry) ReadAccessValidator {
+	return readAccessValidator{tokenRegistry: tr}
+}
+
+// NFTOwnerCanRead checks if the nft owner/peer can read the document
+// Note: signature should be calculated from the hash which is calculated as
+// keccak256("\x19Ethereum Signed Message:\n"${message length}${message}).
+func (r readAccessValidator) NFTOwnerCanRead(
+	cd *coredocumentpb.CoreDocument,
+	registry common.Address,
+	tokenID []byte,
+	signature string,
+	peer identity.CentID) error {
+
+	// check if the peer can read the doc
+	if r.PeerCanRead(cd, peer) {
+		return nil
+	}
+
+	// check if the nft is present in read rules
+	found := findRole(cd, coredocumentpb.Action_ACTION_READ, func(role *coredocumentpb.Role) bool {
+		return isNFTInRole(role, registry, tokenID)
+	})
+
+	if !found {
+		return errors.New("nft missing")
+	}
+
+	// get the owner of the NFT
+	owner, err := r.tokenRegistry.OwnerOf(registry, tokenID)
+	if err != nil {
+		return errors.New("failed to get NFT owner: %v", err)
+	}
+
+	msg, err := constructNFT(registry, tokenID)
+	if err != nil {
+		return err
+	}
+
+	if !secp256k1.VerifySignatureWithAddress(owner.String(), signature, msg) {
```

Comment:
If I understand this correctly, you check that the signature matches the owner.String(). Owner.String is the Ethereum address of the token owner.

What do you sign? (registry+tokenID) 

In the end, the only thing you check here, is that the actual token owner signed a statement (registry+tokenID). Here's a way to exploit it:

Let's assume Alice and Eve collaborate on a document together. 
1) Alice mint's an NFT for Bob. Bob signs the registry+tokenID creating `sig_bob`. 
2) Bob sends Eve a request to read the document with `sig_bob`
3) Alice removes Eve from the collaborators and updates the document.
    *Eve should not have access to the updated document anymore but Bob still has access*
4) Eve can send an NFT access request to Alice and impersonate Bob just by sending `sig_bob` 👿 

The problem with the signature is that it doesn't scope the permission to one key or identity. If we really want to support this access scheme, the signature needs to be scoped to one identity:

Signature = registry + tokenId + identity